snuba/clickhouse/http.py

```diff
@@ -46,44 +46,106 @@ def encode(self, value: WriterTableRow) -> JSONRow:
+
+    def with_format(self, format: str) -> InsertStatement:
+        self.__format = format
+        return self
+
+    def with_columns(self, column_names: Sequence[str]) -> InsertStatement:
+        self.__column_names = column_names
+        return self
+
+    def get_qualified_table(self) -> str:
+        return (
+            f"{self.__database}.{self.__table_name}"
+            if self.__database
+            else self.__table_name
+        )
+
+    def build_statement(self) -> str:
+        columns_statement = (
+            f"({','.join(self.__column_names)}) " if self.__column_names else ""
+        )
+        format_statement = f" FORMAT {self.__format}" if self.__format else ""
+        return f"INSERT INTO {self.__database}.{self.__table_name} {columns_statement} {format_statement}"
+
+
 class HTTPWriteBatch:
+    """
+    Sends a batch of binary data as payload of a Clickhouse insert statement.
+    It takes care of issuing the HTTP request and parsing the response.
+
+    Data is provided in blocks of data (expressed as bytes). Each call to
+    `append` adds a block to the batch.
+    One HTTP request is created at the initialization and remains alive for
+    the lifetime of the object itself.
+    In order to issue a separate INSERT, a new instance has to be created.
+
+    Batches can be broken into chunks by defining the chunk size. This drives
+    the size of each element of the iterable that is provided to the
+    `urlopen` method.
+    Example:
+    Values appended:
+    [ v1, v2, v3, v4]
+    If chunk size is 2, these values will be sent:
+    [v1, v2] [v3, v4]
+
+    Batches can be partially or entirely buffered locally. This behavior
+    is controlled by `buffer_size`.
+    Setting buffer_size to 0 means all the values are enqueued locally
+    while they are sent to the server. `append` would not block in this
+    case.
+    If the buffer size is higher and the buffer is full, the `append`
+    command will block while the value is sent to the server.
+    """
+
     def __init__(
         self,
         executor: ThreadPoolExecutor,
         pool: HTTPConnectionPool,
-        database: str,
-        table_name: str,
         user: str,
         password: str,
+        statement: InsertStatement,
+        encoding: Optional[str],
+        buffer_size: int,  # 0 means unbounded
         options: Mapping[str, Any],  # should be ``Mapping[str, str]``?
         chunk_size: Optional[int] = None,
     ) -> None:
         if chunk_size is None:
             chunk_size = settings.CLICKHOUSE_HTTP_CHUNK_SIZE
 
-        self.__queue: SimpleQueue[Union[JSONRow, None]] = SimpleQueue()
+        self.__queue: Union[
+            Queue[Union[bytes, None]], SimpleQueue[Union[bytes, None]]
+        ] = Queue(buffer_size) if buffer_size else SimpleQueue()
 
         body = self.__read_until_eof()
         if chunk_size > 1:
             body = (b"".join(chunk) for chunk in chunked(body, chunk_size))
         elif not chunk_size > 0:
```

Comment:
I know this isn't related to your change, but it's weird this check happens after we read all the data.

Reply:
are you referring to the `chunk_size > 0` part or to the whole if block ?

Reply:
The chunk_size > 0 (and < 1) check.

Reply:
Well, line 133 cannot happen before as it needs the content read at line 131. I am not sure I am getting the issue.

Reply:
The check at line 134 does not depend on reading the data. It could happen before the data is actually read. Again, it's not a blocking issue.